lib/preboots/scheduler.js

```diff
@@ -0,0 +1,234 @@
+ * @param {Object} options - Configuration object
+ */
+function Scheduler(options) {
+  EE.call(this);
+  this.log = options.log;
+  this.nsq = options.nsq;
+  this.topic = options.topic;
+  this.models = options.models;
+  this.conc = options.concurrency || 5;
+  this.defaultInterval = options.interval || 60 * 60 * 1000;
+
+  this.intervals = new Map();
+}
+
+/**
+ * Set an interval to schedule catch up jobs for the given env
+ * @function setInterval
+ * @param {String} env - Environment
+ * @param {Number} time - milliseconds of interval
+ * @returns {undefined}
+ */
+Scheduler.prototype.setInterval = function interval(env, time) {
+  time = time || this.defaultInterval;
+  const int = setInterval(() => {
+    this.emit('schedule');
+    this.schedule(env, (err, counts) => {
+      if (err) return this.log.error('Failed to schedule jobs for %s', env, {
+        error: err.message,
+        stack: err.stack
+      });
+      this.emit('scheduled', err, counts);
+      this.log.info('Successfully scheduled catch up jobs for %s', env, counts);
+    });
+  }, time);
+
+  this.intervals.set(env, int);
+};
+
+/**
+ * Clear the given setIntervals for the given env or all of them
+ * @function clear
+ * @param {String} env - Optional env to pass
+ * @returns {undefined}
+ */
+Scheduler.prototype.clear = function clear(env) {
+  if (env) return this._clear(env);
+  for (const key of this.intervals.keys()) {
+    this._clear(key);
+  }
+};
+
+/**
+ * Core clearing of setIntervals that cleans up map value as well
+ * @function _clear
+ * @api private
+ * @param {String} env - Environment
+ * @returns {undefined}
+ */
+Scheduler.prototype._clear = function (env) {
+  clearInterval(this.intervals.get(env));
+  this.intervals.delete(env);
+};
+
+/**
+ * Schedule catch up jobs for the given environment over nsq
+ * @function schedule
+ * @api public
+ * @param {String} env - Environment to schedule for
+ * @param {Function} callback - Continuation to call when completed
+ * @returns {undefined}
+ */
+Scheduler.prototype.schedule = function schedule(env, callback) {
+  const counts = {};
+  function done(err) {
+    if (err) return callback(err);
+    callback(null, counts);
+  }
+  //
+  // Scheduling algorithm
+  //
+  // 1. Fetch all packages.
+  // 2. Fetch all build-heads for each package for a given environment
+  // 3. If a build head version is less than any of its peer build heads,
+  //    trigger a build for that given locale
+  //
+  this.packages((err, packages) => {
+    if (err) return callback(err);
+    async.eachLimit(packages, this.conc, (pkg, next) => {
+      const fn = once(next);
+      const name = pkg.name;
+      this.models.BuildHead.findAll({ name, env }, (err, heads) => {
+        if (err) return fn(err);
+        if (!heads || !heads.length) return fn();
+
+        const lookup = new Lookup(pkg, heads);
+        const writer = nsqStream.createWriteStream(this.nsq.writer, this.topic);
+
+        const missing = lookup.missing();
```

Comment:
I'm a bit unsure if the daemon should be doing these lookups. The fetching logic seems familiar (although I haven't looked at wrhs logic in a while) to what the registry is doing? Would it make sense to pass the packages as part of the post body to `/catchup`

Reply:
@Swaagie we are looking at all packages that have been published to warehouse which we assume our under our management control. This could be overridden by part of a request body if we wanted to specify certain ones but this makes it so there isnt any config management in terms of what to build. We just look at what data we have.

Reply:
My $0.02 here is that `carpenterd` is as good a place as any to do this work. If it turns out to be more generally useful we could make it a module, but today this is the correct place from the POV of encapsulation & networking.

- Encapsulation: we don't want folks talking to C* directly, which this logic does. Therefore it has to be in a service. 
- Networking: the network load will be lighter if we have only one writer to NSQ – in this case `carpenterd`.  If we were to go with `wrhs` it would be multiple writers and multiple readers.